waspc/waspls/src/Wasp/LSP/Completions/ExprCompletion.hs

```diff
@@ -0,0 +1,63 @@
+module Wasp.LSP.Completions.ExprCompletion
+  ( getCompletions,
+  )
+where
+
+import Control.Lens ((?~), (^.))
+import Control.Monad.Log.Class (MonadLog (logM))
+import Control.Monad.Reader.Class (MonadReader, asks)
+import Data.Maybe (maybeToList)
+import qualified Data.Text as Text
+import qualified Language.LSP.Types as LSP
+import qualified Language.LSP.Types.Lens as LSP
+import Wasp.Analyzer.Parser.CST (SyntaxNode)
+import qualified Wasp.Analyzer.Parser.CST as S
+import Wasp.Analyzer.Parser.CST.Traverse
+import Wasp.LSP.Completions.Common (CompletionContext, CompletionProvider, makeBasicCompletionItem)
+import qualified Wasp.LSP.Completions.Common as Ctx
+import Wasp.LSP.Syntax (findChild, isAtExprPlace, lexemeAt)
+
+-- | If the location is at an expression, find declaration names in the file
+-- and return them as completion items.
+--
+-- TODO: include completions for enum variants (use standard type defs from waspc)
+getCompletions :: (MonadReader CompletionContext m, MonadLog m) => CompletionProvider m
+getCompletions location =
+  if not (isAtExprPlace location)
+    then do
+      logM "[ExprCompletion] not at expression"
+      return []
+    else do
+      logM "[ExprCompletion] at expression"
+      src <- asks (^. Ctx.src)
+      syntax <- asks (^. Ctx.cst)
+      let declNames = findDeclNames src syntax
+      logM $ "[ExprCompletion] declnames=" ++ show declNames
+      return $
+        map
+          ( \(name, typ) ->
+              makeBasicCompletionItem (Text.pack name)
+                & (LSP.kind ?~ LSP.CiVariable)
+                & (LSP.detail ?~ Text.pack (":: " ++ typ ++ " (declaration type)"))
+          )
+          declNames
+
+-- | Search through the CST and collect all @(declName, declType)@ pairs.
+findDeclNames :: String -> [SyntaxNode] -> [(String, String)]
+findDeclNames src syntax = traverseForDeclNames $ fromSyntaxForest syntax
+  where
+    traverseForDeclNames :: Traversal -> [(String, String)]
+    traverseForDeclNames t = case kindAt t of
+      S.Program -> maybe [] traverseForDeclNames $ down t
```

Comment:
Do we need the app name as completions items? I think we can easily add it later if needed, but now it feels weird getting a completion suggestion with the app name 

<img width="674" alt="Screenshot 2023-06-06 at 11 05 22" src="https://github.com/wasp-lang/wasp/assets/2223680/aafbd76d-906a-43e2-92f3-9781fdd97eb3">

Reply:
@infomiho so we don't yet have smart completion in the sense that it suggests only stuff that matches by type: right now it suggests all identifiers.
While Craig is implementing it now to correctly work for dictionary keys, this will still not make it correctly work for values (by selecting only those that match by type).

Reply:
We can just omit the app declaration type :) that's my suggestion here

Reply:
We can skip this if there is not time now 👍